src/util/any.hpp

```diff
@@ -0,0 +1,269 @@
+    }
+};
+
+namespace impl {
+    template <typename T>
+    using any_cast_remove_qual = typename
+        std::remove_cv<typename std::remove_reference<T>::type>::type;
+
+    template <typename ValueType, typename U=any_cast_remove_qual<ValueType>>
+    using test_any_cast_constructable = typename
+        std::conditional<
+               std::is_constructible<ValueType, const U&>::value
+            && std::is_constructible<ValueType, U&>::value
+            && std::is_constructible<ValueType, U>::value,
+            std::true_type,
+            std::false_type>::type;
+
+    template<typename T>
+    inline T move_else_copy(any_cast_remove_qual<T>* p, std::true_type) {
+        return std::move(*p);
+    }
+
+    template<typename T>
+    inline T move_else_copy(any_cast_remove_qual<T>* p, std::false_type) {
+        return *p;
+    }
+} // namespace impl
+
+// If operand is not a null pointer, and the typeid of the requested T matches
+// that of the contents of operand, a pointer to the value contained by operand,
+// otherwise a null pointer.
+template<class T>
+const T* any_cast(const any* operand) {
+    static_assert(
+        impl::test_any_cast_constructable<T>::value,
+        "any_cast requires ");
+
+    if (operand==nullptr || !operand->has_value()) {
+        return nullptr;
+    }
+
+    if (operand->type() != typeid(T)) {
+        throw bad_any_cast();
+    }
+
+    return operand->unsafe_cast<T>();
+}
+
+// If operand is not a null pointer, and the typeid of the requested T matches
+// that of the contents of operand, a pointer to the value contained by operand,
+// otherwise a null pointer.
+template<class T>
+T* any_cast(any* operand) {
+    static_assert(
+        impl::test_any_cast_constructable<T>::value,
+        "any_cast can't construct the desired type");
+
+    if (operand==nullptr || !operand->has_value()) {
+        return nullptr;
+    }
+
+    if (operand->type() != typeid(T)) {
+        throw bad_any_cast();
+    }
+
+    return operand->unsafe_cast<T>();
+}
+
+template<class T>
+T any_cast(const any& operand) {
+    using U = impl::any_cast_remove_qual<T>;
+    auto ptr = any_cast<U>(&operand);
+    if (ptr==nullptr) {
+        throw bad_any_cast();
+    }
+    return static_cast<T>(*ptr);
+}
+
+template<class T>
+T any_cast(any& operand) {
+    using U = impl::any_cast_remove_qual<T>;
+    auto ptr = any_cast<U>(&operand);
+    if (ptr==nullptr) {
+        throw bad_any_cast();
+    }
+    return static_cast<T>(*ptr);
+}
+
+template<class T>
+T any_cast(any&& operand) {
+    using U = impl::any_cast_remove_qual<T>;
+    using do_move = typename
+        std::conditional<
+            std::is_move_constructible<T>::value
+            & !std::is_lvalue_reference<T>::value,
+            std::true_type, std::false_type>::type;
+
+    auto ptr = any_cast<U>(&operand);
+    if (ptr==nullptr) {
+        throw bad_any_cast();
+    }
+    return impl::move_else_copy<T>(static_cast<T*>(ptr), do_move());
```

Comment:
This looks fairly complicated (in terms of conditional moving etc.). Does `return std::move(*any_cast<T>(&operand));` not work? If not, an explicatory comment would help.